lib/spack/spack/package_base.py

```diff
@@ -1416,12 +1416,13 @@ def extendee_spec(self):
         if not self.extendees:
             return None
 
-        deps = []
-
         # If the extendee is in the spec's deps already, return that.
-        for dep in self.spec.traverse(deptype=("link", "run")):
-            if dep.name in self.extendees:
-                deps.append(dep)
+        deps = [
+            dep
+            for dep in self.spec.dependencies(deptype=("link", "run"))
+            for d, when in self.extendees.values()
+            if dep.satisfies(d) and self.spec.satisfies(when)
```

Comment:
`dep.satisfies(d)` is true by construction, but maybe you added this check defensively?

Reply:
Yes, added because I can imagine future changes that would make it not true by construction -- like if we allow separate concretization of run deps in the future. Agreed that it is not relevant for current specs.